File: scrabble-bot-ts/src/sim.ts
// Local self-play arena + a faithful clone of the official starter bot. Port of
// scrabble-bot/src/sim.rs. Lets the TS port be proven offline with Bun.

import { Brain, type Event, type Player } from "./game";
import type { Lexicon, Play } from "./lexicon";
import { Counts, type Persona, ScoreModel } from "./strategy";
import { BAG_COUNTS, BAG_TOTAL, indexLetter, rewardExact, valueOf } from "./tiles";

export enum AuctionType {
  Vickrey,
  FirstPrice,
}

/** xorshift64* PRNG (deterministic). */
export class Rng {
  private s: bigint;
  constructor(seed: number | bigint) {
    this.s = (BigInt(seed) ^ 0x9e3779b97f4a7c15n) | 1n;
  }
  nextU64(): bigint {
    let x = this.s;
    x ^= x >> 12n;
    x ^= (x << 25n) & 0xffffffffffffffffn;
    x ^= x >> 27n;
    this.s = x & 0xffffffffffffffffn;
    return (this.s * 0x2545f4914f6cdd1dn) & 0xffffffffffffffffn;
  }
  below(n: number): number {
    return Number(this.nextU64() % BigInt(n));
  }
}

function bagFull(): Counts {
  const c = new Counts();
  c.n = Uint8Array.from(BAG_COUNTS);
  c.total = BAG_TOTAL;
  return c;
}

function draw(bag: Counts, rng: Rng): number {
  if (bag.total === 0) return -1;
  let r = rng.below(bag.total);
  for (let i = 0; i < 26; i++) {
    r -= bag.n[i];
    if (r < 0) {
      bag.remove(i);
      return i;
    }
  }
  return -1;
}

export interface Standing {
  botId: number;
  persona: string;
  score: number;
  balance: number;
  words: number;
}

export function runMatch(players: Player[], auctionType: AuctionType, seed: number | bigint): Standing[] {
  const n = players.length;
  const RESERVE = 1;
  const START_BALANCE = 100;
  const STARTING_RACK = 5;
  const rng = new Rng(seed);
  const bag = bagFull();
  const ids = Array.from({ length: n }, (_, i) => i);

  // Deal starting racks privately.
  const startRacks = Array.from({ length: n }, () => new Counts());
  for (let s = 0; s < STARTING_RACK; s++) {
    for (const r of startRacks) {
      const idx = draw(bag, rng);
      if (idx >= 0) r.add(idx);
    }
  }
  const tilesToAuction = bag.total;

  for (let i = 0; i < n; i++) {
    players[i].observe({ t: "MatchStart", me: i, players: ids, startBalance: START_BALANCE, tilesToAuction });
    players[i].observe({ t: "RackSnapshot", rack: startRacks[i], balance: START_BALANCE, score: 0 });
  }

  for (;;) {
    const letterIdx = draw(bag, rng);
    if (letterIdx < 0) break;

    const order: Array<[number, number]> = [];
    for (let i = 0; i < n; i++) {
      const amt = players[i].decideBid(letterIdx);
      if (amt >= RESERVE) order.push([i, amt]);
    }
    order.sort((a, b) => b[1] - a[1] || a[0] - b[0]);

    let winner: number | null = null;
    let paid = 0;
    if (order.length > 0) {
      winner = order[0][0];
      paid =
        auctionType === AuctionType.FirstPrice
          ? order[0][1]
          : order.length >= 2
            ? Math.max(order[1][1], RESERVE)
            : RESERVE;
    }
    const resolved: Event = { t: "AuctionResolved", winner, letterIdx, paid };
    for (const p of players) p.observe(resolved);

    for (let i = 0; i < n; i++) {
      const play = players[i].chooseWord();
      if (play) {
        const ev: Event = { t: "WordPlayed", botId: i, word: play.word, reward: Math.trunc(play.reward) };
        for (const p of players) p.observe(ev);
      }
    }
  }

  // End-of-bag flush.
  let guard = 0;
  for (;;) {
    let progress = false;
    for (let i = 0; i < n; i++) {
      const play = players[i].bestForcedWord();
      if (play) {
        const ev: Event = { t: "WordPlayed", botId: i, word: play.word, reward: Math.trunc(play.reward) };
        for (const p of players) p.observe(ev);
        progress = true;
      }
    }
    if (!progress || ++guard > 100) break;
  }

  const standings: Standing[] = players.map((p) => ({
    botId: p.id(),
    persona: p.name(),
    score: p.score(),
    balance: p.balance(),
    words: p.words(),
  }));
  standings.sort((a, b) => b.score - a.score);
  return standings;
}

export interface TourneyResult {
  name: string;
  matches: number;
  wins: number;
  totalScore: number;
  sumRank: number;
}

export function runTournament(
  bardRoster: Persona[],
  nNaive: number,
  matches: number,
  seed: bigint,
  lex: Lexicon,
): TourneyResult[] {
  const agg = new Map<string, TourneyResult>();
  for (let m = 0; m < matches; m++) {
    const players: Player[] = [];
    bardRoster.forEach((persona, i) => players.push(new Brain(i, lex, persona, ScoreModel.CumulativeWordScore)));
    for (let k = 0; k < nNaive; k++) players.push(new NaiveBot(bardRoster.length + k, lex));

    const matchSeed = (seed + BigInt(m) * 0x9e3779b97f4a7c15n) & 0xffffffffffffffffn;
    const standings = runMatch(players, AuctionType.Vickrey, matchSeed);

    standings.forEach((s, rank) => {
      let e = agg.get(s.persona);
      if (!e) {
        e = { name: s.persona, matches: 0, wins: 0, totalScore: 0, sumRank: 0 };
        agg.set(s.persona, e);
      }
      e.matches++;
      if (rank === 0) e.wins++;
      e.totalScore += s.score;
      e.sumRank += rank;
    });
  }
  return [...agg.values()].sort((a, b) => b.wins - a.wins || b.totalScore - a.totalScore);
}

// --- NaiveBot: faithful clone of bot-starter/src/strategy.ts -----------------

export class NaiveBot implements Player {
  private rack = new Counts();
  private bal = 0;
  private sc = 0;
  private w = 0;
  constructor(
    private meId: number,
    private lex: Lexicon,
  ) {}

  private longestWord(): Play | null {
    let best: { base: number; len: number; letters: Uint8Array } | null = null;
    let bestLen = 1; // min word length is 2
    this.lex.forEachPlay(this.rack.n, (path, depth, base) => {
      if (depth > bestLen && depth <= 15) {
        bestLen = depth;
        best = { base, len: depth, letters: Uint8Array.from(path.subarray(0, depth)) };
      }
    });
    if (!best) return null;
    const b = best as { base: number; len: number; letters: Uint8Array };
    let word = "";
    for (let k = 0; k < b.len; k++) word += indexLetter(b.letters[k]);
    return { word, baseScore: b.base, len: b.len, reward: rewardExact(b.base, b.len) };
  }

  observe(ev: Event): void {
    switch (ev.t) {
      case "MatchStart":
        this.meId = ev.me;
        this.bal = ev.startBalance;
        this.sc = 0;
        this.w = 0;
        this.rack = new Counts();
        break;
      case "RackSnapshot":
        this.rack = ev.rack;
        this.bal = ev.balance;
        this.sc = ev.score;
        break;
      case "AuctionResolved":
        if (ev.winner === this.meId) {
          this.rack.add(ev.letterIdx);
          this.bal -= ev.paid;
        }
        break;
      case "WordPlayed":
        if (ev.botId === this.meId) {
          const leave = this.rack.tryRemoveWord(ev.word);
          if (leave) this.rack = leave;
          this.bal += ev.reward;
          this.sc += ev.reward;
          this.w++;
        }
        break;
      case "Participant":
        if (ev.botId === this.meId) {
          this.bal = ev.balance;
          this.sc = ev.score;
        }
        break;
    }
  }

  decideBid(letterIdx: number): number {
    const value = valueOf(letterIdx);
    const have = this.rack.get(letterIdx);
    const want = have === 0 ? value + 2 : Math.max(1, value);
    return Math.min(want, Math.max(0, this.bal));
  }
  chooseWord(): Play | null {
    return this.longestWord();
  }
  bestForcedWord(): Play | null {
    return this.longestWord();
  }
  id(): number {
    return this.meId;
  }
  name(): string {
    return "Naive (starter)";
  }
  score(): number {
    return this.sc;
  }
  balance(): number {
    return this.bal;
  }
  words(): number {
    return this.w;
  }
}

File: scrabble-bot/src/sim.rs
//! Local self-play arena that mirrors the ScrabbleBot ruleset faithfully enough
//! to (a) prove the engine works end-to-end and (b) tune personas by self-play —
//! all without the live server or credentials.
//!
//! Modelled rules (verbatim from the module): start balance 100, starting rack 5
//! tiles, 98-tile bag (no blanks), Vickrey pricing with reserve 1, ties to the
//! earliest bidder, word reward = `base * num / denom`, **winner = highest
//! cumulative `score`**. It also fields [`NaiveBot`], a faithful clone of the
//! official `bot-starter/src/strategy.ts`, so The Bard can be measured against
//! the baseline everyone else starts from.

use std::collections::BTreeMap;
use std::sync::Arc;

use crate::bidding::ScoreModel;
use crate::dict::{Lexicon, Play};
use crate::game::{Bot, Event, Player};
use crate::persona::Persona;
use crate::rack::Counts;
use crate::tiles::{index_letter, reward_exact, value_of, BAG_COUNTS, BAG_TOTAL, NLETTERS};

/// Auction pricing rule.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum AuctionType {
    Vickrey,
    FirstPrice,
}

/// A small, fast, deterministic PRNG (xorshift64*). Public so the benchmark can
/// reuse it.
pub struct Rng(u64);

impl Rng {
    pub fn new(seed: u64) -> Self {
        Rng((seed ^ 0x9E37_79B9_7F4A_7C15) | 1)
    }
    pub fn next_u64(&mut self) -> u64 {
        let mut x = self.0;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.0 = x;
        x.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }
    pub fn below(&mut self, n: u64) -> u64 {
        self.next_u64() % n
    }
}

fn draw(bag: &mut Counts, rng: &mut Rng) -> Option<usize> {
    if bag.total == 0 {
        return None;
    }
    let mut r = rng.below(bag.total as u64) as i64;
    for i in 0..NLETTERS {
        r -= bag.n[i] as i64;
        if r < 0 {
            bag.remove(i);
            return Some(i);
        }
    }
    None
}

/// Final standing of one participant in a match.
#[derive(Clone, Debug)]
pub struct Standing {
    pub bot_id: u64,
    pub persona: String,
    pub score: i64,
    pub balance: i64,
    pub words: u32,
}

/// Run a single match to completion and return standings sorted by score (desc).
pub fn run_match(
    mut players: Vec<Box<dyn Player>>,
    auction_type: AuctionType,
    seed: u64,
) -> Vec<Standing> {
    let n = players.len();
    const RESERVE: i64 = 1;
    const START_BALANCE: i64 = 100;
    const STARTING_RACK: u32 = 5;

    let mut rng = Rng::new(seed);
    let mut bag = Counts {
        n: BAG_COUNTS,
        total: BAG_TOTAL as u16,
    };
    let ids: Vec<u64> = (0..n as u64).collect();

    // Deal starting racks privately (opponents never learn these — realistic).
    let mut start_racks = vec![Counts::new(); n];
    for _ in 0..STARTING_RACK {
        for r in start_racks.iter_mut() {
            if let Some(idx) = draw(&mut bag, &mut rng) {
                r.add(idx);
            }
        }
    }
    let tiles_to_auction = bag.total as u32;

    for (i, p) in players.iter_mut().enumerate() {
        p.observe(&Event::MatchStart {
            me: i as u64,
            players: ids.clone(),
            start_balance: START_BALANCE,
            tiles_to_auction,
        });
        p.observe(&Event::RackSnapshot {
            rack: start_racks[i],
            balance: START_BALANCE,
            score: 0,
        });
    }

    let mut auction_id = 0u64;
    while let Some(letter_idx) = draw(&mut bag, &mut rng) {
        auction_id += 1;

        // Sealed bids.
        let mut order: Vec<(usize, i64)> = Vec::with_capacity(n);
        for (i, p) in players.iter().enumerate() {
            let amt = p.decide_bid(letter_idx);
            if amt >= RESERVE {
                order.push((i, amt));
            }
        }
        // Highest wins; ties to the earliest (lowest index, mirroring lowest id).
        order.sort_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));
        let (winner, paid) = if order.is_empty() {
            (None, 0)
        } else {
            let (w, top) = order[0];
            let paid = match auction_type {
                AuctionType::FirstPrice => top,
                AuctionType::Vickrey => {
                    if order.len() >= 2 {
                        order[1].1.max(RESERVE)
                    } else {
                        RESERVE
                    }
                }
            };
            (Some(w as u64), paid)
        };

        let ev = Event::AuctionResolved {
            winner,
            letter_idx,
            paid,
        };
        for p in players.iter_mut() {
            p.observe(&ev);
        }

        // Each player may cash one word this round (hold/play decided internally).
        for i in 0..n {
            if let Some(play) = players[i].choose_word() {
                let ev = Event::WordPlayed {
                    bot_id: i as u64,
                    reward: play.reward as i64,
                    word: play.word,
                };
                for p in players.iter_mut() {
                    p.observe(&ev);
                }
            }
        }
        let _ = auction_id;
    }

    // End of bag: force-flush remaining racks — any unplayed tile scores nothing.
    let mut guard = 0;
    loop {
        let mut progress = false;
        for i in 0..n {
            if let Some(play) = players[i].best_forced_word() {
                let ev = Event::WordPlayed {
                    bot_id: i as u64,
                    reward: play.reward as i64,
                    word: play.word,
                };
                for p in players.iter_mut() {
                    p.observe(&ev);
                }
                progress = true;
            }
        }
        guard += 1;
        if !progress || guard > 100 {
            break;
        }
    }

    let mut standings: Vec<Standing> = players
        .iter()
        .map(|p| Standing {
            bot_id: p.id(),
            persona: p.name().to_string(),
            score: p.score(),
            balance: p.balance(),
            words: p.words(),
        })
        .collect();
    standings.sort_by(|a, b| b.score.cmp(&a.score));
    standings
}

/// Aggregated tournament result for one persona (or "Naive" as a class).
#[derive(Clone, Debug)]
pub struct TourneyResult {
    pub name: String,
    pub matches: u32,
    pub wins: u32,
    pub total_score: i64,
    pub sum_rank: u64,
}

impl TourneyResult {
    pub fn win_rate(&self) -> f64 {
        if self.matches == 0 {
            0.0
        } else {
            self.wins as f64 / self.matches as f64
        }
    }
    pub fn avg_score(&self) -> f64 {
        if self.matches == 0 {
            0.0
        } else {
            self.total_score as f64 / self.matches as f64
        }
    }
    pub fn avg_rank(&self) -> f64 {
        if self.matches == 0 {
            0.0
        } else {
            self.sum_rank as f64 / self.matches as f64
        }
    }
}

/// Run `matches` matches of `bard_roster` (each as a Bard `Bot`) plus `n_naive`
/// faithful starter clones, and aggregate by persona name.
pub fn run_tournament(
    bard_roster: &[Persona],
    n_naive: usize,
    matches: u32,
    seed: u64,
    lex: Arc<Lexicon>,
) -> Vec<TourneyResult> {
    let mut agg: BTreeMap<String, TourneyResult> = BTreeMap::new();

    for m in 0..matches {
        let mut players: Vec<Box<dyn Player>> = Vec::new();
        for (i, &persona) in bard_roster.iter().enumerate() {
            players.push(Box::new(Bot::new(
                i as u64,
                persona,
                ScoreModel::default(),
                lex.clone(),
            )));
        }
        let mut idx = bard_roster.len();
        for _ in 0..n_naive {
            players.push(Box::new(NaiveBot::new(idx as u64, lex.clone())));
            idx += 1;
        }

        let match_seed = seed.wrapping_add((m as u64).wrapping_mul(0x9E37_79B9_7F4A_7C15));
        let standings = run_match(players, AuctionType::Vickrey, match_seed);

        for (rank, s) in standings.iter().enumerate() {
            let e = agg.entry(s.persona.clone()).or_insert_with(|| TourneyResult {
                name: s.persona.clone(),
                matches: 0,
                wins: 0,
                total_score: 0,
                sum_rank: 0,
            });
            e.matches += 1;
            if rank == 0 {
                e.wins += 1;
            }
            e.total_score += s.score;
            e.sum_rank += rank as u64;
        }
    }

    let mut v: Vec<TourneyResult> = agg.into_values().collect();
    v.sort_by(|a, b| {
        b.wins
            .cmp(&a.wins)
            .then(b.total_score.cmp(&a.total_score))
    });
    v
}

// ---------------------------------------------------------------------------
// NaiveBot — a faithful clone of bot-starter/src/strategy.ts (the baseline).
// ---------------------------------------------------------------------------

/// Faithful reimplementation of the official starter strategy:
///   * `decideBid`: want = (held? max(1,value) : value+2), clamped to balance.
///   * `chooseWord`: play the single longest formable word, every chance it gets.
pub struct NaiveBot {
    me: u64,
    lex: Arc<Lexicon>,
    rack: Counts,
    balance: i64,
    score: i64,
    words: u32,
}

impl NaiveBot {
    pub fn new(me: u64, lex: Arc<Lexicon>) -> Self {
        Self {
            me,
            lex,
            rack: Counts::new(),
            balance: 0,
            score: 0,
            words: 0,
        }
    }

    fn longest_word(&self) -> Option<Play> {
        let mut best: Option<(u32, usize, [u8; 15])> = None;
        let mut best_len = 1usize; // min word length is 2
        self.lex.for_each_play(&self.rack, |word, base| {
            let len = word.len();
            if len > best_len && len <= 15 {
                best_len = len;
                let mut l = [0u8; 15];
                l[..len].copy_from_slice(word);
                best = Some((base, len, l));
            }
        });
        best.map(|(base, len, l)| {
            let word: String = l[..len].iter().map(|&i| index_letter(i as usize) as char).collect();
            Play {
                word,
                base_score: base,
                len,
                reward: reward_exact(base as i64, len) as f64,
            }
        })
    }
}

impl Player for NaiveBot {
    fn observe(&mut self, ev: &Event) {
        match ev {
            Event::MatchStart { me, start_balance, .. } => {
                self.me = *me;
                self.balance = *start_balance;
                self.score = 0;
                self.words = 0;
                self.rack = Counts::new();
            }
            Event::RackSnapshot { rack, balance, score } => {
                self.rack = *rack;
                self.balance = *balance;
                self.score = *score;
            }
            Event::AuctionResolved { winner, letter_idx, paid } => {
                if *winner == Some(self.me) {
                    self.rack.add(*letter_idx);
                    self.balance -= *paid;
                }
            }
            Event::WordPlayed { bot_id, word, reward } => {
                if *bot_id == self.me {
                    if let Some(leave) = self.rack.try_remove_word(word.as_bytes()) {
                        self.rack = leave;
                    }
                    self.balance += *reward;
                    self.score += *reward;
                    self.words += 1;
                }
            }
            Event::Participant { bot_id, balance, score } => {
                if *bot_id == self.me {
                    self.balance = *balance;
                    self.score = *score;
                }
            }
            Event::AuctionOpen { .. } | Event::MatchEnd => {}
        }
    }

    fn decide_bid(&self, letter_idx: usize) -> i64 {
        let value = value_of(letter_idx) as i64;
        let have = self.rack.get(letter_idx);
        let want = if have == 0 { value + 2 } else { value.max(1) };
        want.min(self.balance.max(0))
    }

    fn choose_word(&self) -> Option<Play> {
        self.longest_word()
    }

    fn best_forced_word(&self) -> Option<Play> {
        self.longest_word()
    }

    fn id(&self) -> u64 {
        self.me
    }
    fn name(&self) -> &str {
        "Naive (starter)"
    }
    fn score(&self) -> i64 {
        self.score
    }
    fn balance(&self) -> i64 {
        self.balance
    }
    fn words(&self) -> u32 {
        self.words
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_match_runs_and_produces_a_ranked_winner() {
        let lex = Arc::new(Lexicon::load_embedded());
        let players: Vec<Box<dyn Player>> = vec![
            Box::new(Bot::new(0, Persona::shakespeare(), ScoreModel::default(), lex.clone())),
            Box::new(NaiveBot::new(1, lex.clone())),
            Box::new(NaiveBot::new(2, lex.clone())),
            Box::new(NaiveBot::new(3, lex.clone())),
        ];
        let standings = run_match(players, AuctionType::Vickrey, 12345);
        assert_eq!(standings.len(), 4);
        // Scores are sorted descending and everyone played at least something.
        assert!(standings[0].score >= standings[3].score);
        let total_words: u32 = standings.iter().map(|s| s.words).sum();
        assert!(total_words > 0, "somebody should have played a word");
    }
}
